src/lib/pubkey/ec_group/curve_gfp.h

```diff
@@ -11,215 +11,73 @@
+      bool operator==(const CurveGFp& other) const { return (m_group == other.m_group); }
 
-      size_t get_p_bytes() const { return (get_p_bits() + 7) / 8; }
-
-      size_t get_ws_size() const { return m_repr->get_ws_size(); }
-
-      const BigInt& get_a_rep() const { return m_repr->get_a_rep(); }
-
-      const BigInt& get_b_rep() const { return m_repr->get_b_rep(); }
-
-      const BigInt& get_1_rep() const { return m_repr->get_1_rep(); }
-
-      bool a_is_minus_3() const { return m_repr->a_is_minus_3(); }
-
-      bool a_is_zero() const { return m_repr->a_is_zero(); }
-
-      bool is_one(const BigInt& x) const { return m_repr->is_one(x); }
-
-      BigInt invert_element(const BigInt& x, secure_vector<word>& ws) const { return m_repr->invert_element(x, ws); }
-
-      void to_rep(BigInt& x, secure_vector<word>& ws) const { m_repr->to_curve_rep(x, ws); }
-
-      void from_rep(BigInt& x, secure_vector<word>& ws) const { m_repr->from_curve_rep(x, ws); }
-
-      BigInt from_rep_to_tmp(const BigInt& x, secure_vector<word>& ws) const {
-         BigInt xt(x);
-         m_repr->from_curve_rep(xt, ws);
-         return xt;
-      }
-
-      // TODO: from_rep taking && ref
-
-      void mul(BigInt& z, const BigInt& x, const BigInt& y, secure_vector<word>& ws) const {
-         m_repr->curve_mul(z, x, y, ws);
-      }
-
-      void mul(BigInt& z, const word x_w[], size_t x_size, const BigInt& y, secure_vector<word>& ws) const {
-         m_repr->curve_mul_words(z, x_w, x_size, y, ws);
-      }
-
-      void sqr(BigInt& z, const BigInt& x, secure_vector<word>& ws) const { m_repr->curve_sqr(z, x, ws); }
-
-      void sqr(BigInt& z, const word x_w[], size_t x_size, secure_vector<word>& ws) const {
-         m_repr->curve_sqr_words(z, x_w, x_size, ws);
-      }
-
-      BigInt mul(const BigInt& x, const BigInt& y, secure_vector<word>& ws) const { return mul_to_tmp(x, y, ws); }
-
-      BigInt sqr(const BigInt& x, secure_vector<word>& ws) const { return sqr_to_tmp(x, ws); }
-
-      BigInt mul_to_tmp(const BigInt& x, const BigInt& y, secure_vector<word>& ws) const {
-         BigInt z;
-         m_repr->curve_mul(z, x, y, ws);
-         return z;
-      }
-
-      BigInt sqr_to_tmp(const BigInt& x, secure_vector<word>& ws) const {
-         BigInt z;
-         m_repr->curve_sqr(z, x, ws);
-         return z;
+   private:
+      const EC_Group_Data& group() const {
+         BOTAN_ASSERT_NONNULL(m_group);
+         return *m_group;
       }
 
-      void swap(CurveGFp& other) { std::swap(m_repr, other.m_repr); }
-
-      friend void swap(CurveGFp& x, CurveGFp& y) { x.swap(y); }
-
       /**
-      * Equality operator
-      * @param other a curve
-      * @return true iff *this is the same as other
+      * Raw pointer
+      *
+      * This EC_Group_Data is not owned because instead the EC_Group_Data
+      * owns this CurveGFp, so we can always access it safely. If it was
+      * a shared_ptr this would cause a reference cycle.
```

Comment:
It would work but it would complicate things because we'd have to pay the weak->shared upgrade cost each time the group is used (eg for each point addition/doubling!). Not a huge fan of raw pointers, for the obvious reasons, but here the potential for misuse is relatively low; the only way a `CurveGFp` can be constructed at all is via a private/friend only constructor, it is only constructed in one place, and that code is careful to maintain the liveness invariant.

If this was something we were stuck with long term I'd probably look for a cleaner solution but this seems to me good enough to work until such time as the whole class can just be deleted.